lib/bundler.rb

```diff
@@ -143,8 +145,41 @@ def ruby_scope
       "#{Bundler.rubygems.ruby_engine}/#{Bundler.rubygems.config_map[:ruby_version]}"
     end
 
+    def user_home
+      @user_home ||= begin
+        home = Bundler.rubygems.user_home
+        warning = "Your home directory is not set properly:"
+        if home.nil?
+          warning += "\n * It is not set at all"
+        elsif !File.directory?(home)
+          warning += "\n * `#{home}` is not a directory"
+        elsif !File.writable?(home)
+          warning += "\n * `#{home}` is not writable"
+        else
+          return @user_home = Pathname.new(home)
+        end
+
+        login = Etc.getlogin || "unknown"
+
+        tmp_home = Pathname.new(Dir.tmpdir).join("bundler", "home", login)
+        begin
+          SharedHelpers.filesystem_access(tmp_home, :write) do |p|
+            FileUtils.mkdir_p(p)
+          end
+        rescue => e
+          warning += "\n\nBundler also failed to create a temporary home directory at `#{tmp_home}`:\n#{e}"
+          raise warning
+        end
+
+        warning += "\n\nBundler will use `#{tmp_home}` as your home directory temporarily"
+
+        Bundler.ui.warn(warning)
+        tmp_home
+      end
+    end
+
     def user_bundle_path
-      Pathname.new(Bundler.rubygems.user_home).join(".bundle")
+      Pathname.new(user_home).join(".bundle")
```

Comment:
`user_name` is `Pathname` already. So it should be safe to leave `user_home.join(".bundle")`